edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,493 @@
+    pub id: Uuid,
+}
+
+let query = "with account := 
+    (insert Account {
+        username := <str>$0
+    }),
+    select <json>account {
+      username,
+      id
+    };";
+
+// We know there will only be one result so use query_single_json;
+// otherwise it will return a map of json
+let json_res = client
+    .query_single_json(query, &("SomeUserName",))
+    .await?
+    .unwrap();
+
+let as_account: Account = serde_json::from_str(&json_res)?;
+```
+
+## Using JSON with the `edgedb(json)` attribute
+
+Adding the `edgedb(json)` attribute on top of `Queryable` allows unpacking a struct from JSON returned from EdgeDB in a single call:
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+#[edgedb(json)]
+pub struct JsonQueryableAccount {
+    pub username: String,
+    pub id: Uuid,
+}
+
+let json_queryable_accounts: Vec<JsonQueryableAccount> = client
+    .query("select <json>Account { username, id }", &())
+    .await
+    .unwrap();
+```
+
+This attribute can also be used on an inner struct value that implements `Queryable`:
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+pub struct InnerJsonQueryableAccount {
+    pub username: String,
+    pub id: Uuid,
+    #[edgedb(json)]
+    pub some_json: HashMap<String, String>,
+}
+``` 
+
+## Transactions
+
+The client also has a `.transaction()` method that allows atomic [transactions](https://www.edgedb.com/docs/edgeql/transactions). Wikipedia has a good example of a transaction and why it would be best done atomically:
+
+```
+An example of an atomic transaction is a monetary transfer from bank account A
+to account B. It consists of two operations, withdrawing the money from 
+account A and saving it to account B. Performing these operations in an atomic
+transaction ensures that the database remains in a consistent state, that is,
+money is neither lost nor created if either of those two operations fails.
+```
+
+A transaction removing 10 cents from one customer's account and placing it in another's would look like this:
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+pub struct BankCustomer {
+    pub name: String,
+    pub bank_balance: i32,
+}
+// Customer1 has an account with 110 cents in it.
+// Customer2 has an account with 90 cents in it.
+// Customer1 is going to send 10 cents to Customer 2. This will be a transaction
+// because we don't want the case to ever occur - even for a split second -
+// where one account has sent money while the other has not received it yet.
+
+// After the transaction is over, each customer should have 100 cents.
+
+let query = "with customer := (
+    update BankCustomer filter .name = <str>$0
+    set { bank_balance -= 10 }
+    ),
+    select customer {
+      name,
+      bank_balance
+    };";
+
+cloned_client
+    .transaction(|mut conn| async move {
+        let _res_1: BankCustomer = conn.query_required_single(query, &(c1,)).await?;
+        let _res_2: BankCustomer = conn.query_required_single(query, &(&c2,)).await?;
```

Comment:
You are subtracting from both accounts.

Also, this is a good example to actually check balance. Either by returning final balance from the first account and throwing exceptions or by doing "select" at the start.

I think select at the start is better because demonstrates logic inside the transaction and demonstrates why we have to retry user function fully instead of just retrying executed queries. And the fact that we retry transactions automatically have to be at least said prominently here or maybe demostrated with some side-effect function log message inside the transaction function.

Also `select customer` makes code more complex than is needed here. I think you can use [`execute`](https://github.com/edgedb/edgedb-rust/pull/247), as it should be merged into the same release as this tutorial.

Reply:
Okay, gave this one a go here in addition to a short section on the new execute method and how to use it for those who are eager: https://github.com/edgedb/edgedb-rust/commit/7fba9d71142aa21f8466bf6f50c85d9a7696274c

Not sure if that's the ideal transaction method yet, what do you think? It does a balance check beforehand and now uses .execute() but can't think of a way to demonstrate inside a short sample how retries get done automatically. Is there a good way to force a query to fail once and have it retry automatically inside a sample meant to be short and readable?

Also the setting a superuser was the best idea I could find for .execute() as a sample that returns an Ok(()) but doesn't make any huge changes like schema or configuration, which are disabled via the client by default anyway. I spent some time looking for a command that doesn't really make any changes or do anything consequential at all (like a ping) but couldn't find one. Maybe there's a more suitable example for the .execute() section that returns an Ok(())?